src/analysis/normed_space/inner_product.lean

```diff
@@ -0,0 +1,1553 @@
+    norm_num [h₁],
+  end,
+  conj_sym :=
+  begin
+    intros x y,
+    unfold inner,
+    rw [←finset.sum_hom finset.univ conj],
+    apply finset.sum_congr rfl,
+    rintros z -,
+    apply inner_conj_sym,
+    apply_instance
+  end,
+  nonneg_im :=
+  begin
+    intro x,
+    unfold inner,
+    rw[←finset.sum_hom finset.univ im],
+    apply finset.sum_eq_zero,
+    rintros z -,
+    exact inner_self_nonneg_im,
+    apply_instance
+  end,
+  add_left := λ x y z,
+    show ∑ i, inner (x i + y i) (z i) = ∑ i, inner (x i) (z i) + ∑ i, inner (y i) (z i),
+    by simp only [inner_add_left, finset.sum_add_distrib],
+  smul_left := λ x y r,
+    show ∑ (i : ι), inner (r • x i) (y i) = (conj r) * ∑ i, inner (x i) (y i),
+    by simp only [finset.mul_sum, inner_smul_left]
+}
+
+/-- A field `K` satisfying `is_R_or_C` is itself a `K`-inner product space. -/
+instance is_R_or_C.inner_product_space : inner_product_space 𝕜 𝕜 :=
+{ inner := (λ x y, (conj x) * y),
+  norm_sq_eq_inner := λ x, by unfold inner; rw [mul_comm, mul_conj, of_real_re, norm_sq, norm_sq_eq_def],
+  conj_sym := λ x y, by simp [mul_comm],
+  nonneg_im := λ x, by rw[mul_im, conj_re, conj_im]; ring,
+  add_left := λ x y z, by simp [inner, add_mul],
+  smul_left := λ x y z, by simp [inner, mul_assoc] }
+
+
+/-- The standard real/complex Euclidean space, functions on a finite type. For an `n`-dimensional space
+use `euclidean_space K (fin n)`.  -/
+@[reducible, nolint unused_arguments]
+def euclidean_space (𝕜 : Type*) [nondiscrete_normed_field 𝕜] [normed_algebra ℝ 𝕜] [is_R_or_C 𝕜]
+  (n : Type*) [fintype n] : Type* := pi_Lp 2 one_le_two (λ (i : n), 𝕜)
+
+section pi_Lp
+local attribute [reducible] pi_Lp
+variables {ι : Type*} [fintype ι]
+
+instance : finite_dimensional 𝕜 (euclidean_space 𝕜 ι) := by apply_instance
+
+@[simp] lemma findim_euclidean_space :
+  finite_dimensional.findim 𝕜 (euclidean_space 𝕜 ι) = fintype.card ι := by simp
+
+lemma findim_euclidean_space_fin {n : ℕ} :
+  finite_dimensional.findim 𝕜 (euclidean_space 𝕜 (fin n)) = n := by simp
+
+end pi_Lp
+
+
+/-! ### Orthogonal projection in inner product spaces -/
+
+section orthogonal
+
+open filter
+
+include 𝕜
+/--
+Existence of minimizers
+Let `u` be a point in an inner product space, and let `K` be a nonempty complete convex subset.
+Then there exists a unique `v` in `K` that minimizes the distance `∥u - v∥` to `u`.
+ -/
+-- FIXME this monolithic proof causes a deterministic timeout with `-T50000`
+-- It should be broken in a sequence of more manageable pieces,
+-- perhaps with individual statements for the three steps below.
+/- Normally we would want `[is_scalar_tower ℝ 𝕜 α]` in applications as well, to ensure that this
+`[normed_space ℝ α]` instance is compatible with the `[normed_space 𝕜 α]` defined above, but
+here this is not strictly needed as these lemmas never use the 𝕜 instance. -/
+theorem exists_norm_eq_infi_of_complete_convex [normed_space ℝ α] {K : set α} (ne : K.nonempty) (h₁ : is_complete K)
```

Comment:
You shouldn't need `normed_space ℝ α` here. Instead, you should have an instance from `normed_space k α` to `normed_space ℝ α`. Or, better, just formulate the theorem for real vector spaces, and let `restrict_scalars` automatically handle the complex case for you (you would need an instance that a complex inner product space is also a real inner product space, which is good to have in any case)

Reply:
I actually ran into annoying problems when trying to define these kinds of instances (where `is_R_or_C` implies real). If I try to declare, for example, that `semimodule 𝕜 α` implies `semimodule ℝ α`, Lean complains about the case where `𝕜 = ℝ`, because then this new instance is not defeq to the original one. I have no idea how to solve this -- if you have any ideas that would be great!

Reply:
What about just having an instance that `inner_product_space C alpha` implies `inner_product_space R alpha`? And stating all the projection theorems only in real inner product spaces. Then they would also become automatically available in complex inner product spaces as well (since they are just a statement about the norm), which should be enough for applications.

Reply:
I added the instance (it certainly can't hurt!), but I think it would still be good to prove these for the `is_R_or_C` case directly, because once we split into two cases, we're condemned to also split everything that gets built on top...

Reply:
I don't agree with you here: this projection business is specifically a real thing (it requires convexity of the set, which only makes sense in a real context), so it should only be stated for real vector spaces.

The version that would make sense for any `is_R_or_C` field `𝕜` is that, for any `𝕜`-subspace, there is a unique projection on `𝕜` minimizing the distance (which can be stated using only the distance, so not mentioning reals at all). And if you can find properties on the scalar product that only mention `𝕜`, then they are worth stating in this context.

So, my suggestion is: 
* Do all the projection business with convex subsets, and everything that mentions `ℝ`, in a specific section that does not mention any field `𝕜`
* Then deduce the results that make sense over any `is_R_or_C` field (and that don't mention `ℝ` at all) in another mention, by registering locally an instance that the space is also a real inner product space, and using the previous results for real inner product spaces. This instance would be a bad instance as you mention (as for `𝕜 = ℝ` is creates non-defeq instances) but if it is just a local instance specific to these proofs this is not an issue.

If you don't see how to do this, I can have a go at it if you like.

Reply:
Ah I see -- alright I'll give it a shot!

Reply:
This isn't going too well. That instance does really strange things, even locally (`abel` stops working in random places, the typeclass system goes nuts on things like `semiring ℝ` in some of these statements, etc). Now I just rewrote that whole section to just be for real inner products and left the instance unused -- feel free to have a shot at it though.

Reply:
The trick is to *not* register this as an instance (even not a local one), but only to use it in proofs to declare explicitly the real structure on a given space, with `letI`. This means that Lean will not try the instance on this real structure, so nothing bad will happen. 

I have done this, and everything went pretty well! Now, the only thing that is specific to real spaces is the projection on convex sets. Orthogonal projection on complete subspaces is done in the general setting of an `is_R_or_C` field (and I had nothing to change to the proofs, so it definitely convinced me that this is a really efficient way to do things at the same time on reals and complexes!)

It looks ready for merging to me, but before I can merge it someone should have a look to the changes I have made in my last commit. Can you check that you're happy with what I've done?

Reply:
I'm certainly happy with it for what it's worth -- thanks a lot!